### ISSUE TITLE
Problem with adding support at location with decimal values

### ISSUE THREAD

**Issue by sparky1911** [sympy/sympy#21660]
from sympy import *
from sympy.physics.continuum_mechanics.beam import Beam
E, I = symbols('E, I')
b = Beam(9, E, I)
b.apply_load(-12, 9, -1)
b.apply_load(-50, 5, -2)
b.apply_load(-8, 0, 0, end=5)
b.apply_support(0, type='fixed')
b.apply_support(4.5, type='roller')
b.load

after adding the above load i am getting the following equation
𝑀0⟨𝑥⟩−2+𝑅0⟨𝑥⟩−1+𝑅4.50000000000000⟨𝑥−4.5⟩−1−8⟨𝑥⟩0−50⟨𝑥−5⟩−2+8⟨𝑥−5⟩0−12⟨𝑥−9⟩−1

how i am supposed to find the reaction forces with those trailing zeros?
please help

**Comment by oscarbenjamin** [comment]
I don't understand what the issue is. You can use `Rational(9, 2)` to have an exact result here.

**Comment by sparky1911** [comment]
youe see those reaction forces with the trailing zeros to calculate 

`n=2
listreac = b.applied_loads[-n:]
reac = []

for a_tuple in listreac:
    reac.append(a_tuple[0])

r1=str(reac[0])
r2=str(reac[1])
R_0=symbols(r1)
R_4 = symbols(r2)
b.solve_for_reaction_loads(R_4,R_0)
b.reaction_loads`

i need to write the above code to calculate reactions without typing it out manually.
is there a way that it can be automated?

**Comment by oscarbenjamin** [comment]
> i need to write the above code to calculate reactions without typing it out manually.
> is there a way that it can be automated?

Put the code in a function and call the function?

I'm sorry but you need to explain yourself more clearly. I don't understand what this issue is about.

**Comment by sparky1911** [comment]
uh so when you use the apply_support() function with an decimal value as the location is assigns a symbol  and then you need to assign this symbol to a variable like this R= symbols('R4.50000000000000') 
and and then you can use the  solve_for_reaction_loads() to get the reactions loads but is there a way to automate the above steps ? i am trying to write a function but i am getting "IndexError: tuple index out of range" error
i am trying to make a tkinter app so trying to mimimize users inputs.

**Comment by connorferster** [comment]
@oscarbenjamin and @sparky1911, 

I am seeing the issue here more generally. To me, it seems that the need enter symbols as arguments into `.solve_for_reaction_loads()` seems perhaps redundant. By entering in the reactions as unknowns (e.g. `R1`, `R2`) into the `Beam` object, it should already know what the unknowns are without the user having to enter them again separately. In @sparky1911's case, the symbol being generated is difficult to recreate in a separate process because it is something like `R4.500000000000000`. The actual line of code in `beam.py`: 

```python
solution = list((linsolve([shear_curve, moment_curve] + slope_eqs
    837                             + deflection_eqs, (C3, C4) + reactions).args)[0])
```

I think this line seems to be trying to do too much in one expression and, as a result, it is difficult to know what is causing errors and so a user just sees `IndexError: tuple index out of range` but the user does not know if they have entered in too many symbols, incorrect symbols, or something else.

I have a similar use case in mind as @sparky1911 and would be interested in having the `Beam.solve_for_reaction_loads()` method have an update in one of the following ways: 

1. Take symbols as optional arguments
2. Break out line 836 with some `try...except...` blocks so it can do some error reporting
3. Remove arguments from this method altogether

Am interested in hearing thoughts of others on this, e.g. is there something I am missing in my assessment?

**Comment by sparky1911** [comment]
@connorferster  yeah exactly the same problem you are talking about passing the arguments in Beam.solve_for_reaction_loads() seems redundant to me.I am actually trying to develop and API which would return the sfd and bmd of the beam and dont want to expose the end user to any of the code.

and when the load locations are in decimal values creating dummy variables for them is kinda tricky for me.Still looking for a solution.

**Comment by smichr** [comment]
This question was answered in the mailing list [here](https://groups.google.com/g/sympy/c/SRj433Q9Xb4/m/eS9feAeUBwAJ)
```python
>>> eq = b.load
>>> r = sorted([i for i in eq.free_symbols if i.name.startswith("R")], key=lambda x: float(x.name[2:]))
>>> r
[R_0, R_4.50000000000000]
```
Now refer to them as `r[0]` and `r[1]` or put them in a dict with the float value as the key and recall them by float:
```python
>>> R = dict([(float(i.name[2:]), i) for i in r])
>>> R[4.5]
R_4.50000000000000
```

**Comment by connorferster** [comment]
Hi @smichr!

This is similar to the solution I came up with last night: 

```python
reactions = list({b.load.free_symbols} - {b.variable})
```
To state it more clearly, it seems unnecessary to have to provide the reaction symbols to `Beam.solve_for_reaction_loads()`. Given the above code snippet, it seems like this line could be added to the method and then `Beam.solve_for_reaction_loads()` could be called without arguments.

When solving beams, the analyst puts in all of the variables, loads, and boundary conditions into the original statement of the problem. Once the problem is set up, we would expect to be able to just say something like, "ok, solve the beam". I think for this reason, it could make sense to include the expression for retrieving the reactions from the problem statement within `Beam.solve_for_reaction_loads()` so it could be called as a simple "solve" method.

Let me know your thoughts on this. If you think it makes sense, I would be happy to submit a PR updating the method, any tests, and the documentation.

**Comment by smichr** [comment]
I am always in favor of reducing redundant input. And allowing for multiple symbols (and supplying all if none are given) is a common idiom in SymPy.

Would it make sense to simply number the reaction loads instead of labelling them with the load value? At least trailing 0s could be removed. And perhaps better -- use Rationals for them so you get 9/2 instead of 4.50000000000.
